### PR TITLE
Here is implemented 

### PR DESCRIPTION
**Pull request by sevenlays**
[DEMO LINK](https://sevenlays.github.io/react_todo-app-loading-todos/)

- [x] - implement todos loading 
- [x] - implement error messages
- [x] - implement filtering by status

### CODE REVIEW

File: src/components/Footer.tsx [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import { Todo } from '../types/Todo';
+import React, { SetStateAction, Dispatch } from 'react';
+
+type FooterProps = {
+  todos: Todo[];
+  filterType: FilterType;
+  setFilter: Dispatch<SetStateAction<FilterType>>;
+};
+
+const filtersButtons = ['All', 'Active', 'Completed'];
```
Comment (denyssheremeta) [inline src/components/Footer.tsx:11]: If you use enum to type filter and preparing filter section to scaling you can use this:
```
export enum FilterType {
  All = 'All',
  Active = 'Active',
  Completed = 'Completed',
}

const filterValues = Object.values(FilterType);
```
just make array of type/values from your enum type, and map in footer

**Review by denyssheremeta — APPROVED** [review]
Good job! Thank you! Check small issue!

File: src/components/Footer.tsx (hunk 1 above)
Comment (sevenlays) [reply src/components/Footer.tsx:11]: Oh, thats super simple, thank you!

File: src/App.tsx [hunk 2]
```diff
@@ -1,204 +1,136 @@
...
+
+  const handleDeleteTodo = (todoId: number) => {
+    setTodoLoadingId(todoId);
+    setTimeout(() => {
+      deleteTodo(todoId)
+        .catch(() => {
+          triggerError('deleteTodoError');
+        })
+        .finally(() => setTodoLoadingId(null));
+      setTodos(currentTodos =>
```
Comment (mrkovaliv) [inline src/App.tsx:49]: move this callback to separate func

File: src/App.tsx [hunk 3]
```diff
@@ -1,204 +1,136 @@
...
+
+  const handleChangeTodo = (e: ChangeEvent<HTMLInputElement>) => {
+    setTodo({ ...todo, title: e.target.value });
+  };
+
+  const handleUpdateTodo = (updatedTodo: Todo) => {
+    setTodoLoadingId(updatedTodo.id);
+    updateTodo(updatedTodo)
+      .then(todoUpdated => {
+        setTodos(currentTodos => {
```
Comment (mrkovaliv) [inline src/App.tsx:63]: move this callback to separate func

File: src/App.tsx [hunk 4]
```diff
@@ -1,204 +1,136 @@
...
-          {/* this button should be disabled if there are no completed todos */}
-          <button
-            type="button"
-            className="todoapp__clear-completed"
-            data-cy="ClearCompletedButton"
-          >
-            Clear completed
-          </button>
-        </footer>
+        {todos.length !== 0 && (
```
Comment (mrkovaliv) [inline src/App.tsx:126]: just `{todos.length && (`

**Review by mrkovaliv — CHANGES_REQUESTED** [review]
Nice work 👍

File: src/hooks/useTodos.ts [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+    if (!todo.title.trim()) {
+      triggerError('titleError');
+
+      return;
+    }
+
+    // const tempId = Date.now();
+    // const tempTodoItem = { ...todo, id: tempId };
+
+    // setTodos(prev => [...prev, tempTodoItem]);
```
Comment (DanilWeda) [inline src/hooks/useTodos.ts:91]: Remove commented. If you need that, leave a comment that you need it in the next iteration.

File: src/hooks/useTodos.ts [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+        } catch {
+          triggerError('loadingError');
+        } finally {
+          setIsLoading(false);
+        }
+      }
+    };
+
+    fetchTodos();
+  }, []);
```
Comment (DanilWeda) [inline src/hooks/useTodos.ts:73]: You can disabled this rule in this case

File: src/components/SingleTodo.tsx [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+          </button>
+        </>
+      )}
+
+      <TodoLoader todo={todo} />
+    </div>
+  );
+};
+
+// {/* This todo is an active todo */}
```
Comment (DanilWeda) [inline src/components/SingleTodo.tsx:106]: Remove commented. If you need that, leave a comment that you need it in the next iteration.

File: src/components/Footer.tsx [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+export const Footer: React.FC<FooterProps> = ({
+  todos,
+  filterType,
+  setFilter,
+}) => {
+  const activeTodos = todos.filter(t => !t.completed);
```
Comment (DanilWeda) [inline src/components/Footer.tsx:18]: One-letter arguments in function are not a good idea. Please make it more readable.

File: src/components/Footer.tsx (hunk 1 above)
Comment (DanilWeda) [reply src/components/Footer.tsx:11]: Also made these changes.

File: src/components/ErrorsContainer.tsx [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+      {/* show only one message at a time */}
+      {errors.loadingError && 'Unable to load todos'}
+      {/* <br /> */}
+      {errors.titleError && 'Title should not be empty'}
+      {/* <br /> */}
+      {errors.addTodoError && 'Unable to add a todo'}
+      {/* <br /> */}
+      {errors.deleteTodoError && 'Unable to delete a todo'}
+      {/* <br /> */}
+      {errors.updatedTodo && 'Unable to update a todo'}
```
Comment (DanilWeda) [inline src/components/ErrorsContainer.tsx:41]: A better way to move all these messages in constants.

File: src/App.tsx [hunk 10]
```diff
@@ -1,204 +1,79 @@
...
+          newTodo={todo}
+          submit={handleSubmit}
+          changeTodo={handleChangeTodo}
+        />
+        <Main
+          filteredTodos={filteredTodos}
+          deleteTodo={handleDeleteTodo}
+          updateTodo={handleUpdateTodo}
+        />
+        {todos.length !== 0 && (
```
Comment (DanilWeda) [inline src/App.tsx:72]:
```suggestion
        {!!todos.length && (
```
as alternative

File: src/App.tsx [hunk 11]
```diff
@@ -1,204 +1,79 @@
...
+import { Todo } from './types/Todo';
+import { Main } from './components/Main';
+import { useTodosContext } from './Context/TodosContext';
+
+const tempTodo: Todo = {
+  id: 0,
+  userId: USER_ID,
+  title: '',
+  completed: false,
+};
```
Comment (DanilWeda) [inline src/App.tsx:25]: Create separate file for this defaultTodo end import, then use in component

**Pull request closed without merging**